In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/cars-dataset/2. Cars Data1.csv


In [2]:
import pandas as pd

In [3]:
car = pd.read_csv('../input/cars-dataset/2. Cars Data1.csv')

In [4]:
car

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         23.0   
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
427  $38,203          2.4        5.0       197.0      21.0         28.0   
428  $40,083          2.3        5.0       242.0      20.0         26.0   
429  $42,573          2.9        6.0       268.0      19.0         26.0   
430  $24,641          1.9        4.0       170.0      22.0         29.0   
431  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0  

[432 rows x 15 columns]

In [5]:
car.tail()

Make                    Model   Type  Origin DriveTrain      MSRP  \
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
427  $38,203          2.4        5.0       197.0      21.0         28.0   
428  $40,083          2.3        5.0       242.0      20.0         26.0   
429  $42,573          2.9        6.0       268.0      19.0         26.0   
430  $24,641          1.9        4.0       170.0      22.0         29.0   
431  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0

In [6]:
car.shape

(432, 15)

432 rows and 15 columns

### CLEANING

In [7]:
## remove NULL

car.isnull()

Make  Model   Type  Origin  DriveTrain   MSRP  Invoice  EngineSize  \
0    False  False  False   False       False  False    False       False   
1    False  False  False   False       False  False    False       False   
2    False  False  False   False       False  False    False       False   
3    False  False  False   False       False  False    False       False   
4    False  False  False   False       False  False    False       False   
..     ...    ...    ...     ...         ...    ...      ...         ...   
427  False  False  False   False       False  False    False       False   
428  False  False  False   False       False  False    False       False   
429  False  False  False   False       False  False    False       False   
430  False  False  False   False       False  False    False       False   
431  False  False  False   False       False  False    False       False   

     Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  Wheelbase  Length  
0        False       False     False        False   False      False   False  
1        False       False     False        False   False      False   False  
2        False       False     False        False   False      False   False  
3        False       False     False        False   False      False   False  
4        False       False     False        False   False      False   False  
..         ...         ...       ...          ...     ...        ...     ...  
427      False       False     False        False   False      False   False  
428      False       False     False        False   False      False   False  
429      False       False     False        False   False      False   False  
430      False       False     False        False   False      False   False  
431      False       False     False        False   False      False   False  

[432 rows x 15 columns]

In [8]:
car.isnull().sum()

Make           4
Model          4
Type           4
Origin         4
DriveTrain     4
MSRP           4
Invoice        4
EngineSize     4
Cylinders      6
Horsepower     4
MPG_City       4
MPG_Highway    4
Weight         4
Wheelbase      4
Length         4
dtype: int64

In [9]:
car.notnull()

Make  Model  Type  Origin  DriveTrain  MSRP  Invoice  EngineSize  \
0    True   True  True    True        True  True     True        True   
1    True   True  True    True        True  True     True        True   
2    True   True  True    True        True  True     True        True   
3    True   True  True    True        True  True     True        True   
4    True   True  True    True        True  True     True        True   
..    ...    ...   ...     ...         ...   ...      ...         ...   
427  True   True  True    True        True  True     True        True   
428  True   True  True    True        True  True     True        True   
429  True   True  True    True        True  True     True        True   
430  True   True  True    True        True  True     True        True   
431  True   True  True    True        True  True     True        True   

     Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  Wheelbase  Length  
0         True        True      True         True    True       True    True  
1         True        True      True         True    True       True    True  
2         True        True      True         True    True       True    True  
3         True        True      True         True    True       True    True  
4         True        True      True         True    True       True    True  
..         ...         ...       ...          ...     ...        ...     ...  
427       True        True      True         True    True       True    True  
428       True        True      True         True    True       True    True  
429       True        True      True         True    True       True    True  
430       True        True      True         True    True       True    True  
431       True        True      True         True    True       True    True  

[432 rows x 15 columns]

In [10]:
car.notnull().sum()

Make           428
Model          428
Type           428
Origin         428
DriveTrain     428
MSRP           428
Invoice        428
EngineSize     428
Cylinders      426
Horsepower     428
MPG_City       428
MPG_Highway    428
Weight         428
Wheelbase      428
Length         428
dtype: int64

Total rows= 432

sum of null 4 values for each column


and for notnull function : 428 rows have no null value

So, this table has 4 rows with collect null value in it

In [11]:
## find null value in any column then replace it with the mean of that column

car['Cylinders'].fillna(car['Cylinders'].mean(), inplace=True)

In [12]:
car.isnull().sum()

Make           4
Model          4
Type           4
Origin         4
DriveTrain     4
MSRP           4
Invoice        4
EngineSize     4
Cylinders      0
Horsepower     4
MPG_City       4
MPG_Highway    4
Weight         4
Wheelbase      4
Length         4
dtype: int64

Cylinders column : no null

In [13]:
car.head(5)

Make           Model   Type Origin DriveTrain      MSRP   Invoice  \
0  Acura             MDX    SUV   Asia        All  $36,945   $33,337    
1  Acura  RSX Type S 2dr  Sedan   Asia      Front  $23,820   $21,761    
2  Acura         TSX 4dr  Sedan   Asia      Front  $26,990   $24,647    
3  Acura          TL 4dr  Sedan   Asia      Front  $33,195   $30,299    
4  Acura      3.5 RL 4dr  Sedan   Asia      Front  $43,755   $39,014    

   EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0         3.5        6.0       265.0      17.0         23.0  4451.0   
1         2.0        4.0       200.0      24.0         31.0  2778.0   
2         2.4        4.0       200.0      22.0         29.0  3230.0   
3         3.2        6.0       270.0      20.0         28.0  3575.0   
4         3.5        6.0       225.0      18.0         24.0  3880.0   

   Wheelbase  Length  
0      106.0   189.0  
1      101.0   172.0  
2      105.0   183.0  
3      108.0   186.0  
4      115.0   197.0

In [14]:
car['Make'].fillna(car['Make'], inplace=True);  ## not int : cannot calculate MEAN
car['Model'].fillna(car['Model'], inplace=True);
car['Type'].fillna(car['Type'], inplace=True);
car['DriveTrain'].fillna(car['DriveTrain'], inplace=True);
car['MSRP'].fillna(car['MSRP'], inplace=True);
car['Invoice'].fillna(car['Invoice'], inplace=True);
car['EngineSize'].fillna(car['EngineSize'], inplace=True);
car['Horsepower'].fillna(car['Horsepower'], inplace=True);
car['MPG_City'].fillna(car['MPG_City'], inplace=True);
car['MPG_Highway'].fillna(car['MPG_Highway'], inplace=True);
car['Weight'].fillna(car['Weight'], inplace=True);
car['Wheelbase'].fillna(car['Wheelbase'], inplace=True);
car['Length'].fillna(car['Length'], inplace=True);

In [15]:
car.isnull().sum()

Make           4
Model          4
Type           4
Origin         4
DriveTrain     4
MSRP           4
Invoice        4
EngineSize     4
Cylinders      0
Horsepower     4
MPG_City       4
MPG_Highway    4
Weight         4
Wheelbase      4
Length         4
dtype: int64

In [16]:
## count discrete of each manufacturer

car['Make'].value_counts()

Toyota           28
Chevrolet        27
Mercedes-Benz    26
Ford             23
BMW              20
Audi             19
Honda            17
Nissan           17
Volkswagen       15
Chrysler         15
Dodge            13
Mitsubishi       13
Volvo            12
Jaguar           12
Hyundai          12
Subaru           11
Pontiac          11
Mazda            11
Lexus            11
Kia              11
Buick             9
Mercury           9
Lincoln           9
Saturn            8
Cadillac          8
Suzuki            8
Infiniti          8
GMC               8
Acura             7
Porsche           7
Saab              7
Land Rover        3
Oldsmobile        3
Jeep              3
Scion             2
Isuzu             2
MINI              2
Hummer            1
Name: Make, dtype: int64

In [17]:
## origin Asia or Europe

car[car['Origin'].isin(['Asia','Europe'])]

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         23.0   
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
427  $38,203          2.4        5.0       197.0      21.0         28.0   
428  $40,083          2.3        5.0       242.0      20.0         26.0   
429  $42,573          2.9        6.0       268.0      19.0         26.0   
430  $24,641          1.9        4.0       170.0      22.0         29.0   
431  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0  

[281 rows x 15 columns]

In [18]:
## remove records which show weight >4000
car[car['Weight'] >4000]

Make                           Model    Type  Origin DriveTrain  \
0         Acura                             MDX     SUV    Asia        All   
15         Audi  A4 3.0 Quattro convertible 2dr   Sedan  Europe        All   
17         Audi              A6 4.2 Quattro 4dr   Sedan  Europe        All   
18         Audi                A8 L Quattro 4dr   Sedan  Europe        All   
20         Audi                        RS 6 4dr  Sports  Europe      Front   
..          ...                             ...     ...     ...        ...   
405  Volkswagen                      Touareg V6     SUV  Europe        All   
415  Volkswagen                     Phaeton 4dr   Sedan  Europe      Front   
416  Volkswagen                 Phaeton W12 4dr   Sedan  Europe      Front   
419  Volkswagen                       Passat W8   Wagon  Europe      Front   
420       Volvo                         XC90 T6     SUV  Europe        All   

         MSRP   Invoice  EngineSize  Cylinders  Horsepower  MPG_City  \
0    $36,945   $33,337          3.5        6.0       265.0      17.0   
15   $44,240   $40,075          3.0        6.0       220.0      18.0   
17   $49,690   $44,936          4.2        8.0       300.0      17.0   
18   $69,190   $64,740          4.2        8.0       330.0      17.0   
20   $84,600   $76,417          4.2        8.0       450.0      15.0   
..        ...       ...         ...        ...         ...       ...   
405  $35,515   $32,243          3.2        6.0       220.0      15.0   
415  $65,000   $59,912          4.2        8.0       335.0      16.0   
416  $75,000   $69,130          6.0       12.0       420.0      12.0   
419  $40,235   $36,956          4.0        8.0       270.0      18.0   
420  $41,250   $38,851          2.9        6.0       268.0      15.0   

     MPG_Highway  Weight  Wheelbase  Length  
0           23.0  4451.0      106.0   189.0  
15          25.0  4013.0      105.0   180.0  
17          24.0  4024.0      109.0   193.0  
18          24.0  4399.0      121.0   204.0  
20          22.0  4024.0      109.0   191.0  
..           ...     ...        ...     ...  
405         20.0  5086.0      112.0   187.0  
415         22.0  5194.0      118.0   204.0  
416         19.0  5399.0      118.0   204.0  
419         25.0  4067.0      106.0   184.0  
420         20.0  4638.0      113.0   189.0  

[103 rows x 15 columns]

### REVOVE

In [19]:
## remove
car[~(car['Weight'] >4000)]

Make                    Model   Type  Origin DriveTrain      MSRP  \
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
5    Acura  3.5 RL w/Navigation 4dr  Sedan    Asia      Front  $46,100    
..     ...                      ...    ...     ...        ...       ...   
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
5    $41,100          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
427  $38,203          2.4        5.0       197.0      21.0         28.0   
428  $40,083          2.3        5.0       242.0      20.0         26.0   
429  $42,573          2.9        6.0       268.0      19.0         26.0   
430  $24,641          1.9        4.0       170.0      22.0         29.0   
431  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
5    3893.0      115.0   197.0  
..      ...        ...     ...  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0  

[329 rows x 15 columns]

### INCREASE VALUE

In [20]:
car.head(10)

Make                    Model    Type  Origin DriveTrain      MSRP  \
0  Acura                      MDX     SUV    Asia        All  $36,945    
1  Acura           RSX Type S 2dr   Sedan    Asia      Front  $23,820    
2  Acura                  TSX 4dr   Sedan    Asia      Front  $26,990    
3  Acura                   TL 4dr   Sedan    Asia      Front  $33,195    
4  Acura               3.5 RL 4dr   Sedan    Asia      Front  $43,755    
5  Acura  3.5 RL w/Navigation 4dr   Sedan    Asia      Front  $46,100    
6  Acura   NSX coupe 2dr manual S  Sports    Asia       Rear  $89,765    
7   Audi              A4 1.8T 4dr   Sedan  Europe      Front  $25,940    
8   Audi   A41.8T convertible 2dr   Sedan  Europe      Front  $35,940    
9   Audi               A4 3.0 4dr   Sedan  Europe      Front  $31,840    

    Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0  $33,337          3.5        6.0       265.0      17.0         23.0  4451.0   
1  $21,761          2.0        4.0       200.0      24.0         31.0  2778.0   
2  $24,647          2.4        4.0       200.0      22.0         29.0  3230.0   
3  $30,299          3.2        6.0       270.0      20.0         28.0  3575.0   
4  $39,014          3.5        6.0       225.0      18.0         24.0  3880.0   
5  $41,100          3.5        6.0       225.0      18.0         24.0  3893.0   
6  $79,978          3.2        6.0       290.0      17.0         24.0  3153.0   
7  $23,508          1.8        4.0       170.0      22.0         31.0  3252.0   
8  $32,506          1.8        4.0       170.0      23.0         30.0  3638.0   
9  $28,846          3.0        6.0       220.0      20.0         28.0  3462.0   

   Wheelbase  Length  
0      106.0   189.0  
1      101.0   172.0  
2      105.0   183.0  
3      108.0   186.0  
4      115.0   197.0  
5      115.0   197.0  
6      100.0   174.0  
7      104.0   179.0  
8      105.0   180.0  
9      104.0   179.0

In [21]:
## increase all the values of 'MPG_City' by 3

car['MPG_City'] = car['MPG_City'].apply(lambda x:x+3)

In [22]:
car.head(10)

Make                    Model    Type  Origin DriveTrain      MSRP  \
0  Acura                      MDX     SUV    Asia        All  $36,945    
1  Acura           RSX Type S 2dr   Sedan    Asia      Front  $23,820    
2  Acura                  TSX 4dr   Sedan    Asia      Front  $26,990    
3  Acura                   TL 4dr   Sedan    Asia      Front  $33,195    
4  Acura               3.5 RL 4dr   Sedan    Asia      Front  $43,755    
5  Acura  3.5 RL w/Navigation 4dr   Sedan    Asia      Front  $46,100    
6  Acura   NSX coupe 2dr manual S  Sports    Asia       Rear  $89,765    
7   Audi              A4 1.8T 4dr   Sedan  Europe      Front  $25,940    
8   Audi   A41.8T convertible 2dr   Sedan  Europe      Front  $35,940    
9   Audi               A4 3.0 4dr   Sedan  Europe      Front  $31,840    

    Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0  $33,337          3.5        6.0       265.0      20.0         23.0  4451.0   
1  $21,761          2.0        4.0       200.0      27.0         31.0  2778.0   
2  $24,647          2.4        4.0       200.0      25.0         29.0  3230.0   
3  $30,299          3.2        6.0       270.0      23.0         28.0  3575.0   
4  $39,014          3.5        6.0       225.0      21.0         24.0  3880.0   
5  $41,100          3.5        6.0       225.0      21.0         24.0  3893.0   
6  $79,978          3.2        6.0       290.0      20.0         24.0  3153.0   
7  $23,508          1.8        4.0       170.0      25.0         31.0  3252.0   
8  $32,506          1.8        4.0       170.0      26.0         30.0  3638.0   
9  $28,846          3.0        6.0       220.0      23.0         28.0  3462.0   

   Wheelbase  Length  
0      106.0   189.0  
1      101.0   172.0  
2      105.0   183.0  
3      108.0   186.0  
4      115.0   197.0  
5      115.0   197.0  
6      100.0   174.0  
7      104.0   179.0  
8      105.0   180.0  
9      104.0   179.0